onnxruntime/core/providers/xnnpack/nn/conv.cc

```diff
@@ -0,0 +1,230 @@
+
+  // have to delay creating the xnnpack kernel until after the weights are pre-packed.
+}
+
+// use PrePack to handle the weight layout change as that's not a simple NCHW -> NHWC transpose
+Status Conv::PrePack(const Tensor& tensor, int input_idx, AllocatorPtr alloc,
+                     /*out*/ bool& is_packed,
+                     /*out*/ PrePackedWeights* /*prepacked_weights*/) {
+  is_packed = false;
+
+  // only layout of weight input is adjusted via PrePack
+  if (input_idx == 1) {
+    // Transpose from {M, C/group, kH, kW} to {M, kH, kW, C/group}
+    auto orig_shape = tensor.Shape();
+
+    std::vector<size_t> perm{0, 2, 3, 1};
+    std::vector<int64_t> new_dims{orig_shape[0],
+                                  orig_shape[2],
+                                  orig_shape[3],
+                                  orig_shape[1]};
+
+    packed_w_ = Tensor::Create(tensor.DataType(), TensorShape(new_dims), alloc);
+
+    SingleAxisTranspose(perm, tensor, *packed_w_, /*from*/ 1, /*to*/ 3);
+
+    is_packed = true;
+
+    // we can create the kernel now
+    struct xnn_operator* p = nullptr;
+    ORT_RETURN_IF_ERROR(CreateXnnpackKernel(conv_attrs_, C_, M_, kernel_shape_, clip_min_max_, IsDepthwise(),
+                                            *packed_w_, B_ ? B_->Data<float>() : nullptr, p));
+
+    op0_.reset(p);
+  }
+
+  return Status::OK();
+}
+
+Status Conv::Compute(OpKernelContext* context) const {
+  const Tensor& X = *context->Input<Tensor>(0);  // this is in NHWC format
+  const auto& X_shape = X.Shape();
+  const int64_t N = X_shape[0];  // input is NHWC
+  const int64_t H = X_shape[1];
+  const int64_t W = X_shape[2];
+
+  // We don't need to call ValidateInputShape as we checked validity in ConvChecker.
+  // We also can't use ValidateInputShape as-is as the weight tensor was pre-packed and the layout was changed there.
+  // ORT_RETURN_IF_ERROR(conv_attrs_.ValidateInputShape(&X, &W));
+
+  // CPU Conv starts with TensorShapeVector Y_dims({N, M}); and passes in X->Shape().Slice(2);
+  // We know this is 2D in NHWC format so we need to start with 'N', pass in the H, W, and append M last
+  TensorShapeVector Y_dims({N});
+  TensorShape input_shape = {H, W};
+
+  ConvAttributes::ConvPadVector pads(conv_attrs_.pads);
+  ORT_RETURN_IF_ERROR(conv_attrs_.InferPadsAndOutputShape(input_shape, kernel_shape_,
+                                                          conv_attrs_.strides, conv_attrs_.dilations, pads,
+                                                          Y_dims));
+
+  Y_dims.push_back(M_);
+  Tensor* Y = context->Output(0, TensorShape(Y_dims));
+
+  // Bail out early if one of the dimensions is zero.
+  if (Y->Shape().Size() == 0) {
+    return Status::OK();
+  }
+
+  xnn_status status = xnn_setup_convolution2d_nhwc_f32(op0_.get(), N, H, W, X.Data<float>(), Y->MutableData<float>(),
+                                                       nullptr /*threadpool*/);  // TBD: how to handle threading
```

Comment:
"how to handle threading?" -
With the EP approach maybe we could add a interface as:

ThreadPoolInferface* EP::GetThreadPool(){...}

Framework could then query an EP and set EP thread pool as the intra_op_thread_pool.
For xnnpack ep, it needs to wrapped up pthreadpool by ThreadPoolInterface where some refactorings are needed for  the purpose.